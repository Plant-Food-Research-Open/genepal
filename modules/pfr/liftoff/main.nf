process LIFTOFF {
    tag "$meta.id"
    label 'process_high'

    conda "bioconda::liftoff=1.6.3"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/liftoff:1.6.3--pyhdfd78af_0':
        'biocontainers/liftoff:1.6.3--pyhdfd78af_0' }"

    input:
    tuple val(meta), path(target_fa)
    path ref_fa, name: 'liftoff_reference_assembly.fa' // To avoid name collisions betwen target_fa and ref_fa
    path ref_annotation
    
    output:
    tuple val(meta), path("*.gff3")             , emit: gff3
    tuple val(meta), path("*.polished.gff3")    , emit: polished_gff3, optional: true
    tuple val(meta), path("*.unmapped.txt")     , emit: unmapped_txt
    path "versions.yml"                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    liftoff \\
        -g $ref_annotation \\
        -p $task.cpus \\
        -o "${prefix}.gff3" \\
        -u "${prefix}.unmapped.txt" \\
        $args \\
        $target_fa \\
        liftoff_reference_assembly.fa

    mv "${prefix}.gff3_polished" "${prefix}.polished.gff3" \\
        || echo "-polish is absent"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        liftoff: \$(liftoff --version 2> /dev/null)
    END_VERSIONS
    """
    
    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch "${prefix}.gff3"
    touch "${prefix}.unmapped.txt"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        liftoff: \$(liftoff --version 2> /dev/null)
    END_VERSIONS
    """
}
