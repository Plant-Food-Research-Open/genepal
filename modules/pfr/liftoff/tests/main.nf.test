nextflow_process {

    name "Test Process LIFTOFF"
    script "../main.nf"
    process "LIFTOFF"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "liftoff"

    test("homo_sapiens-genome_21_fasta-genome_1_fasta-genome_1_gtf") {

        setup {
            run("GUNZIP") {
                script "../../../nf-core/gunzip"

                process {
                    """
                    input[0] = [
                        [ id:'test' ],
                        file(params.test_data['homo_sapiens']['genome']['genome_1_fasta'], checkIfExists: true)
                    ]
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.test_data['homo_sapiens']['genome']['genome_21_fasta'], checkIfExists: true)
                ]
                input[1] = GUNZIP.out.gunzip.map { meta, file -> file }
                input[2] = [
                    file(params.test_data['homo_sapiens']['genome']['genome_1_gtf'], checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.unmapped_txt).match("unmapped_txt") },
                { assert snapshot(process.out.versions).match("versions") },
                { assert path(process.out.gff3.get(0).get(1)).getText().contains("chr21\tLiftoff\texon\t34608061\t34608118\t.\t+\t.") },
                { assert path(process.out.polished_gff3.get(0).get(1)).getText().contains("chr21\tLiftoff\texon\t34608061\t34608118\t.\t+\t.") }
            )
        }

    }

    test("stub") {
        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.test_data['homo_sapiens']['genome']['genome_21_fasta'], checkIfExists: true)
                ]
                input[1] = [
                    file(params.test_data['homo_sapiens']['genome']['genome_1_fasta'], checkIfExists: true)
                ]
                input[2] = [
                    file(params.test_data['homo_sapiens']['genome']['genome_1_gtf'], checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.gff3 != null },
                { assert process.out.polished_gff3 == [] },
                { assert process.out.unmapped_txt != null },
                { assert process.out.versions != null },
                { assert snapshot(process.out.versions).match("versions") }
            )
        }

    }

}
