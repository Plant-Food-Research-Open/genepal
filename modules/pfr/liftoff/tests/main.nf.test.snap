{
    "unmapped_txt": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.unmapped.txt:md5,7391d10df6e15db356b084c9af5259e4"
                ]
            ]
        ],
        "timestamp": "2023-12-01T13:57:40.748507"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,205d9c609e7fe27d8199550d842bdce8"
            ]
        ],
        "timestamp": "2023-12-01T13:57:40.752414"
    }
}